bitcoin/src/blockdata/transaction.rs

```diff
@@ -526,6 +534,20 @@ impl TxOut {
         Weight::from_vb(self.size() as u64).expect("should never happen under normal conditions")
     }
 
+    /// Returns the effective_value `Amount` returning `None` on overflow.
+    ///
+    /// The effective value is the value of this output minus the amount to spend it.  That is,
+    /// the effective_value can be calculated as value - (fee_rate * weight).
+    pub fn effective_value(
+        &self,
+        fee_rate: FeeRate,
+        satisfaction_weight: Weight,
+    ) -> Option<Amount> {
+        let weight = satisfaction_weight.checked_add(TxIn::base_weight())?;
+        let input_fee = fee_rate.checked_mul_by_weight(weight)?;
+        self.value.checked_sub(input_fee)
```

Comment:
This should be saturating sub so that the effective amount is zero.

Reply:
Hmm shouldn't this return None if there's an arithmetic error instead of silently succeeding and returning 0?

Reply:
If the value you want to spend is less than fee then the effective value is logically zero. `None` for all other cases.

Reply:
If value is less than the fee, then the effective_value is negative, and you think it should be zero instead of negative, else if there is an overflow/underflow return none.  In either case, the utxo will be filtered out and not used since it's considered a dust transaction.

Reply:
I agree that it should be negative and I think any "effective value" calculation should return a signed amount.

Reply:
> I agree that it should be negative and I think any "effective value" calculation should return a signed amount.

Good idea, I think it makes sense to change this to SignedAmount and return a negative value even though a TxOut is Amount and can't actually be negative.

Reply:
@apoelstra I think zero is more practical since the purpose is to be used in coin selection. The exact negative value doesn't matter there since the input will never be selected. Also I think it kinda doesn't make sense for it to be negative since you're not paying for it in the first place.

Reply:
@Kixunil it could be selected if the user forces it, for example. Or if it has a locktime condition that you don't want to become active (if it has negative value, maybe you don't care...but if it's only a little negative and might become positive, maybe you wanna sweep it just be sure). Or if you are just sweeping the whole wallet.

And if it does get selected, then it needs to have a negative value so that you properly account for it.

Reply:
This now returns a SignedAmount that may be negative, although I feel a bit bad about ignoring the possible error returned by `to_signed()`.  At least this preserves more information than just returning zero.

Reply:
When summing values, it might be nice to add `to_amount()` method so instead of:

```
if effective_value.is_positive() {
    value += Amount::from_sat(effective_value.to_sat() as u64)
}
```

It can just be:
```
if effective_value.is_positive() {
    value += effective_value.to_amount()
}
```

side not that `into_ok()` still is unstable, otherwise this would be alright
```
value += if effective_value.is_positive() {
    effective_value.into_ok()
} else {
    Amount::ZERO
};

```

Reply:
@apoelstra OK then, I'm fine with signed.

@yancyribbens  Good idea to have some handy conversion but should return `Result` or `Option` - ideally using `TryFrom`. And maybe have `saturating_to_unsigned()` on `SignedAmount` to convert negative values to zero.